### PR TITLE
Update animation.js to add interpolation types

### LINKED ISSUE #168: [gltf] Interpolation example not interpolating
Interpolation example from gltf.js file not working
image
https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/InterpolationTest

### PR DESCRIPTION
**Pull request by simonharrisco**
- Hand written so handle with care
- Lots of lines could be saved by addition of vec4.scale or quat.scale
- Used the equation here: 
https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-c-spline-interpolation

I should also note that i know it shouldnt be merged in this starting form up its more a proposed structure (that needs work) and some functioning maths as a foundation

Fix #168

### CODE REVIEW

**Comment by dmnsgn** [comment]
We haven't merged prettier in pex-renderer yet so you can revert your latest commit.

File: animation.js [hunk 1]
```diff
@@ -85,15 +89,128 @@ Animation.prototype.update = function () {
...
+          currentOutput = outputData[prevIndex]
+          break;
+        case 'CUBICSPLINE':
+          //intangent = index
+          //position = index+1
+          //outtangent = index+2
+              
+          let prevInTangent,prevOutTangent,prevPosition,nextInTangent,nextOutTangent,nextPos;
+
+          if(path === 'rotation') {
```
Comment (dmnsgn) [inline animation.js]: Can you make a ternary of this? Knowing that rotation is the only specific case eg `let prevInTangent  = path === 'rotation' ? vec4.create() : vec3.create()`

File: animation.js [hunk 2]
```diff
@@ -85,15 +89,128 @@ Animation.prototype.update = function () {
...
+            ]
+            currentOutput = quat.normalize(tempOutputTest);
+
+          } else {
+            let p0Calc = vec3.scale(prevPosition,((2*ttt) - (3*tt) + 1))
+            let m0Calc = vec3.scale(prevOutTangent,(ttt - (2*tt)+t))
+            let p1Calc = vec3.scale(nextPos,((-2*ttt) + (3*tt)))
+            let m1Calc = vec3.scale(nextInTangent,(ttt-tt))
+  
+            let tempOutputTest = vec3.add(vec3.add(vec3.add(p0Calc,m0Calc),p1Calc),m1Calc)
```
Comment (dmnsgn) [inline animation.js]: What is this tempOutputTest?

File: animation.js [hunk 3]
```diff
@@ -117,6 +234,7 @@ Animation.prototype.update = function () {
   })
 }
 
+
```
Comment (dmnsgn) [inline animation.js]: Try to remove extra space when committing, it makes the process of reviewing leaner.

File: animation.js (hunk 1 above)
Comment (dmnsgn) [reply animation.js]: Actually the ternary could happen below: `let prevOutTangent = outputData[(prevIndex * 3) +2].slice()  : [0, 0, 0]`.

Also quick style comment: add space before and after `*` and `+`

**Comment by vorg** [comment]
What are those actually Vec4 or Quaternions?

```javascript
if (path === 'rotation') {
  prevInTangent = vec4.create(); 
  prevOutTangent = vec4.create(); 
  prevPosition = vec4.create(); 
  nextInTangent = vec4.create();
  nextOutTangent = vec4.create(); 
  nextPos = vec4.create();        
```

**Comment by simonharrisco** [comment]
Thats the thing right, I thought they would be quaternions but the equation has you scale them, which doesn't make sense for a quaternion (right?!) 

So I just used vec4 thinking I could use vec4.scale
 (I mean I now realise I cant, but as I wrote it that was the logic)

**Comment by vorg** [comment]
It would be also good to have a look at garbage collection if `target.transform.set({ position: currentOutput })` doesn't retain the vec3 then we could have preallocated temp array instead of new one every frame.

Stuff like
```javascript
currentOutput = [1,1,1,1]
currentOutput = quat.copy(prevOutput)
```

**Comment by vorg** [comment]
@simonharrisco can link the interpolation docs?

**Comment by simonharrisco** [comment]
https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-c-spline-interpolation

**Comment by simonharrisco** [comment]
` the resulting p(t) quaternion` So it should be quat

**Comment by simonharrisco** [comment]
There is a mistake 
slomo
shouldnt move backwards.

File: animation.js [hunk 4]
```diff
@@ -119,4 +206,4 @@ Animation.prototype.update = function () {
 
 module.exports = function createMorph (opts) {
   return new Animation(opts)
-}
+}
```
Comment (dmnsgn) [inline animation.js]: Doesn't your editor have a final new line setting?

File: animation.js [hunk 5]
```diff
@@ -15,6 +17,8 @@ function Animation (opts) {
   this.prevTime = Date.now() // ms
   this.channels = opts.channels || []
   this.changed = new Signal()
+  this.currentOutputVec3 = vec3.create()
```
Comment (dmnsgn) [inline animation.js]: Instead of attaching it via this, I would extract it at the top of the file as a let because it is only used for the updates..

**Comment by dmnsgn** [comment]
Pending pex-math vec4.scale to be released.

**Pull request merged**